File: packages/mcp-server/src/tools/audit.ts
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { z } from 'zod';
import { pmApi, PmApiError } from '../pm-api.js';

function ok(data: unknown) {
  return { content: [{ type: 'text' as const, text: JSON.stringify(data) }] };
}

function err(message: string) {
  return { content: [{ type: 'text' as const, text: `Error: ${message}` }] };
}

export function registerAuditTools(server: McpServer) {
  server.tool(
    'list_audit_logs',
    'List permission audit logs (member_added, member_removed, role_changed). Admin/owner only.',
    {
      limit: z.number().int().min(1).max(100).optional().describe('Max records to return (default 50)'),
      cursor: z.string().optional().describe('Pagination cursor — created_at of last record (ISO timestamp)'),
    },
    async ({ limit, cursor }) => {
      try {
        const params = new URLSearchParams();
        if (limit !== undefined) params.set('limit', String(limit));
        if (cursor !== undefined) params.set('cursor', cursor);
        const qs = params.size > 0 ? `?${params.toString()}` : '';
        const data = await pmApi(`/api/audit-logs${qs}`);
        return ok(data);
      } catch (e) {
        return err(e instanceof PmApiError ? e.message : String(e));
      }
    },
  );
}

File: backend/sprintable_mcp/tools/audit.py
"""감사 로그 MCP 도구 (1개)."""
from __future__ import annotations

from mcp.types import TextContent

from ..api_client import client
from ..response import err, ok
from ..schemas import SprintableInput


class ListAuditLogsInput(SprintableInput):
    limit: int | None = None
    cursor: str | None = None


async def list_audit_logs(args: ListAuditLogsInput) -> list[TextContent]:
    """권한 감사 로그 조회 (Admin/Owner 전용)."""
    params: dict = {}
    if args.limit is not None:
        params["limit"] = str(args.limit)
    if args.cursor:
        params["cursor"] = args.cursor
    try:
        return ok(await client.get("/api/v2/audit-logs", params=params))
    except Exception as exc:
        return err(str(exc))
